lib/vector-core/src/sink.rs

```diff
@@ -60,6 +76,96 @@ impl fmt::Debug for VectorSink {
 // === StreamSink ===
 
 #[async_trait]
-pub trait StreamSink {
-    async fn run(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()>;
+pub trait StreamSink<T> {
+    async fn run(self: Box<Self>, input: stream::BoxStream<'_, T>) -> Result<(), ()>;
+}
+
+/// Wrapper for sinks implementing `Sink<Event>` to implement
+/// `Sink<EventArray>`. This stores an iterator over the incoming
+/// `EventArray` to be pushed into the wrapped sink one at a time.
+struct EventSink<S> {
+    sink: S,
+    queue: Option<<EventArray as EventContainer>::IntoIter>,
+}
+
+macro_rules! poll_ready_ok {
+    ( $e:expr ) => {
+        match $e {
+            r @ (Poll::Pending | Poll::Ready(Err(_))) => return r,
+            Poll::Ready(Ok(ok)) => ok,
+        }
+    };
+}
+
+impl<S: Sink<Event> + Send + Unpin> EventSink<S> {
+    fn new(sink: S) -> Self {
+        Self { sink, queue: None }
+    }
+
+    fn next_event(&mut self) -> Option<Event> {
+        match &mut self.queue {
+            #[allow(clippy::single_match_else)] // No, clippy, this isn't a single pattern
+            Some(queue) => match queue.next() {
+                Some(event) => Some(event),
+                None => {
+                    // Reset the queue to empty after the last event
+                    self.queue = None;
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+
+    fn flush_queue(self: &mut Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
+        while self.queue.is_some() {
+            poll_ready_ok!(self.sink.poll_ready_unpin(cx));
+            let event = match self.next_event() {
+                None => break,
+                Some(event) => event,
+            };
+            if let Err(err) = self.sink.start_send_unpin(event) {
+                return Poll::Ready(Err(err));
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<S: Sink<Event> + Send + Unpin> Sink<EventArray> for EventSink<S> {
+    type Error = S::Error;
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        poll_ready_ok!(self.flush_queue(cx));
+        self.sink.poll_ready_unpin(cx)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, events: EventArray) -> Result<(), Self::Error> {
+        assert!(self.queue.is_none()); // Should be guaranteed by `poll_ready`
+        self.queue = Some(events.into_events());
+        self.next_event()
+            .map_or(Ok(()), |event| self.sink.start_send_unpin(event))
```

Comment:
It seems like we're adding potentially multiple events to the queue here but only attempting to send one of them to the inner sink's `start_send`. Is that correct? I see the complexity with having reserved one slot by the inner call to `poll_ready` in our own `poll_ready`, so it wouldn't be as simple as calling `flush_queue`. It seems like they'll be flushed fine by the subsequent call to `poll_ready` or `poll_flush`, but I'm not sure there's a good reason to leave them sitting in the meantime (other than additional complexity if it works fine in practice).